### DISCUSSION TITLE
`HeartBeat too slow` problems

### DISCUSSION THREAD

**Discussion by WeihanLi** [dotnet/aspnetcore#28687] [General]
Recently, I got some the `HeartBeat too slow` warning and I can not see any thread starvation, the thread starvation data is from the thread pool etw event data collected by [prometheus-net.DotNetRuntime](https://github.com/djluck/prometheus-net.DotNetRuntime)

The metrics data:

![](https://user-images.githubusercontent.com/7604648/102328655-4cd27b80-3fc2-11eb-9bc0-00bbbcae13c6.png)

The warning log:

![image](https://user-images.githubusercontent.com/7604648/102328867-9622cb00-3fc2-11eb-88c9-1632bfe55f6c.png)

There's a web API project running on asp.net core 3.1 and it's running in the docker container, the docker image I used is `mcr.microsoft.com/dotnet/core/aspnet:3.1-alpine`

**Comment by Tratcher** [comment]
When you see this happening you'll want to capture a dump and check for blocked threads.

Threadpool exhaustion isn't the only cause of this issue, but it is by far the most common.

**Reply by WeihanLi** [reply]
Thanks for the info @Tratcher , but it's a little difficult for me to create a dump in time in production, is there any other possible causes?

I can see there's some thread timeout for thread pool adjustments, is there some known reason for the thread timeout adjustment?

**Reply by Tratcher** [reply]
> I can see there's some thread timeout for thread pool adjustments, is there some known reason for the thread timeout adjustment?

I don't know what that means, can you share more details?

**Reply by WeihanLi** [reply]
I mean there's some [thread pool worker thread adjustment event](https://docs.microsoft.com/en-us/dotnet/fundamentals/diagnostics/runtime-thread-events#threadpoolworkerthreadadjustmentadjustment-event) caused by `Thread timed out` around the time of `HeartBeat too slow` warning

**Reply by Tratcher** [reply]
Ok, what were the details given with that event? Especially the "Reason".

**Reply by WeihanLi** [reply]
I can see there's two reasons for the thread adjustment, `Climbing move` and `Thread timed out` only, No more details.

I'm not sure if the metrics dashboard would help, maybe not

![](https://user-images.githubusercontent.com/7604648/103596338-a5999200-4f38-11eb-89a3-829b74b50b33.png)

**Reply by Tratcher** [reply]
As I understand it, climbing_move is the precursor to full starvation. It's actively injecting new threads into the pool because the system is under pressure. Your system seems to be doing that continuously which suggests you are experiencing starvation inducing conditions. Figuring out what's consuming all of those threads is best done with a dump. I'm not aware of good alternatives.

**Reply by WeihanLi** [reply]
Great thanks for your information.
